atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraRequestExceptionHandler.java

```diff
@@ -0,0 +1,169 @@
+        this.blacklist = blacklist;
+    }
+
+    <K extends Exception> void handleRequest(
+                RetryableCassandraRequest<?, K> req,
+                InetSocketAddress hostTried,
+                Exception ex)
+            throws K {
+        req.incrementNumberOfAttempts();
+        req.triedOnHost(hostTried);
+        this.handleExceptionInternal(req, hostTried, ex);
+        if (isRetriableWithBackoffException(ex)) {
+            // And value between -500 and +500ms to backoff to better spread load on failover
+            int sleepDuration =
+                    req.getNumberOfAttempts() * 1000 + (ThreadLocalRandom.current().nextInt(1000) - 500);
+            log.warn("Retrying a query, {}, with backoff of {}ms, intended for host {}.",
+                    UnsafeArg.of("queryString", req.getFunction().toString()),
+                    SafeArg.of("sleepDuration", sleepDuration),
+                    SafeArg.of("hostName", CassandraLogHelper.host(hostTried)));
+
+            try {
+                Thread.sleep(sleepDuration);
+            } catch (InterruptedException i) {
+                throw new RuntimeException(i);
+            }
+            if (req.getNumberOfAttempts() >= maxTriesSameHost.get()) {
+                req.giveUpOnPreferredHost();
+            }
+        } else if (isFastFailoverException(ex)) {
+            log.info("Retrying with fast failover a query intended for host {}.",
+                    SafeArg.of("hostName", CassandraLogHelper.host(hostTried)));
+            req.giveUpOnPreferredHost();
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private <K extends Exception> void handleExceptionInternal(
+            RetryableCassandraRequest<?, K> req, InetSocketAddress host, Exception ex) throws K {
+        if (isRetriableException(ex) || isRetriableWithBackoffException(ex) || isFastFailoverException(ex)) {
+            if (req.getNumberOfAttempts() >= maxTriesTotal.get()) {
+                if (ex instanceof TTransportException
+                        && ex.getCause() != null
+                        && (ex.getCause().getClass() == SocketException.class)) {
+                    log.error(CONNECTION_FAILURE_MSG, req.getNumberOfAttempts(), ex);
+                    String errorMsg =
+                            MessageFormatter.format(CONNECTION_FAILURE_MSG, req.getNumberOfAttempts()).getMessage();
+                    throw (K) new TTransportException(((TTransportException) ex).getType(), errorMsg, ex);
+                } else {
+                    log.error("Tried to connect to cassandra {} times.",
+                            SafeArg.of("numTries", req.getNumberOfAttempts()), ex);
+                    throw (K) ex;
+                }
+            } else {
+                // Only log the actual exception the first time
+                if (req.getNumberOfAttempts() > 1) {
+                    log.warn("Error occurred talking to cassandra. Attempt {} of {}. Exception message was: {} : {}",
```

Comment:
These are bumped down to INFO in #2776; we should replicate that here. You won't get the merge conflicts until you try to put this into develop.

Reply:
To be specific, I mean the logs on lines 101, 108 and 114.

Reply:
I think you meant 68, 108 and 114. Done.